src/gasman.c

```diff
@@ -2081,54 +2081,68 @@ void CheckMasterPointers( void )
 }
 
 
-/****************************************************************************
-**
-*F  SwapMasterPoint( <bag1>, <bag2> ) . . . swap pointer of <bag1> and <bag2>
-*/
-void SwapMasterPoint (
-    Bag                 bag1,
-    Bag                 bag2 )
+// Swap the master pointers of bag1 and bag2
+//
+// We need to make sure the correct bags are garbage collected, so we always put
+// *both* bags on the ChangedBags linked-list, rather than pick through the
+// exact cases, as it is never incorrect to mark something changed.
+//
+// For completeness and future reference here are the necessary points to
+// consider.
+//
+// As ChangedBags is a singly linked list, we can't easily take anything out of
+// this list. Also, as the link pointers are stored inside the bag itself, when
+// we swap PTR_BAG, we must swap LINK_BAG as well.
+//
+// If both bags are old, and the original bag1 was changed, we need to now mark
```

Comment:
I'd still change "bag1 was changed" to e.g. "bag1 had been marked as changed" or "bag1 was on the changed list" or so